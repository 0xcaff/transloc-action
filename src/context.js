// @flow
import type { Stop } from "transloc-api";
import type { DialogflowApp } from "actions-on-google";
import type { Result } from "./result";
import { getStopById } from "./resolve";

export opaque type LocationKey = string;

export const FROM_STOP_KEY: LocationKey = "from";
export const TO_STOP_KEY: LocationKey = "to";

export type StopParams = {
  stopId: number
};

export const storeLocationContext = (
  app: DialogflowApp,
  key: LocationKey,
  stop: Stop
): void => {
  const lifetime = undefined;

  const params: StopParams = { stopId: stop.id };

  app.setContext(key, lifetime, params);
};

export const getStopFromContext = (
  app: DialogflowApp,
  key: LocationKey,
  stops: Stop[]
): Result<Stop> => {
  const context = app.getContext(key);
  if (!context) {
    return { type: "EMPTY" };
  }

  const params: StopParams = context.parameters;
  const maybeStop: ?Stop = getStopById(params.stopId, stops);
  if (!maybeStop) {
    app.tell("Something went wrong.");
    return { type: "DELEGATING" };
  }

  return { type: "SUCCESS", value: maybeStop };
};
