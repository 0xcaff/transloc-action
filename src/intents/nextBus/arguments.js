// @flow
import type { DialogflowApp } from "actions-on-google";

export const FROM_ARGUMENT = "from";

export const getFromArg = (app: DialogflowApp): ?string =>
  app.getArgument(FROM_ARGUMENT);

export const TO_ARGUMENT = "to";

export const getToArg = (app: DialogflowApp): ?string =>
  app.getArgument(TO_ARGUMENT);
