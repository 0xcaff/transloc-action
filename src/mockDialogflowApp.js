// @flow
import type {
  Response,
  DeviceLocation,
  SupportedPermission,
  SurfaceCapability,
  SimpleResponse,
  Context
} from "actions-on-google";
import {
  RichResponse,
  List,
  OptionItem
} from "actions-on-google/response-builder";
import type { UserData } from "./userData";

type RecordedPermission = {
  context: string,
  permission: SupportedPermission,
  dialogState?: Object
};

type RecordedResponse =
  | { type: "TELL", response: Response }
  | { type: "ASK_FOR_PERMISSION", permission: RecordedPermission }
  | { type: "ASK_WITH_LIST", list: List };

// A mock DialogflowApplication. It is used to pass in arguments and collect.
export class MockDialogflowApp {
  // Output Field
  response: ?RecordedResponse = null;
  contextOut: Map<string, Context<Object>> = new Map();

  // Input Fields
  args: Map<string, any>;
  permissionGranted: boolean = false;
  deviceLocation: ?DeviceLocation;
  selectedOption: ?string;
  surfaceCapabilities: Set<SurfaceCapability> = new Set();
  context: Map<string, Context<Object>> = new Map();
  userStorage: UserData = {
    agency_id: 643
  };

  constructor(args: Map<string, any> = new Map()) {
    this.args = args;
  }

  tell = (speechResponse: Response): ?Object => {
    if (this.response) {
      throw new TypeError("Multiple responses received.");
    }

    this.response = { type: "TELL", response: speechResponse };
  };

  getArgument = (argName: string): any => this.args.get(argName);

  isPermissionGranted = (): boolean => this.permissionGranted;

  askForPermission = (
    context: string,
    permission: SupportedPermission,
    dialogState?: Object
  ): ?Object => {
    if (this.response) {
      throw new TypeError("Multiple responses received.");
    }

    this.response = {
      type: "ASK_FOR_PERMISSION",
      permission: { context, permission, dialogState }
    };
  };

  askWithList = (
    inputPrompt: string | RichResponse | SimpleResponse,
    list: List
  ): Object => {
    if (this.response) {
      throw new TypeError(`Multiple responses received.`);
    }

    this.response = {
      type: "ASK_WITH_LIST",
      list
    };

    return {};
  };

  buildRichResponse = (otherResponse: ?RichResponse): RichResponse => {
    return new RichResponse(otherResponse);
  };

  buildList = (title: string): List => new List(title);

  buildOptionItem = (
    key: ?string,
    synonyms: string | string[] | void
  ): OptionItem => {
    const optionItem = new OptionItem();
    if (key) {
      optionItem.setKey(key);
    }
    if (synonyms) {
      optionItem.addSynonyms(synonyms);
    }
    return optionItem;
  };

  getDeviceLocation = (): ?DeviceLocation => this.deviceLocation;

  getSelectedOption = (): ?string => this.selectedOption;

  hasSurfaceCapability = (cap: SurfaceCapability): boolean =>
    this.surfaceCapabilities.has(cap);

  getContext = (name: string) => this.context.get(name);

  setContext = (name: string, lifespan: ?number, params: Object) => {
    this.contextOut.set(name, {
      name,
      lifespan: lifespan || 0,
      parameters: params
    });
  };

  SupportedPermissions = {
    NAME: "NAME",
    DEVICE_PRECISE_LOCATION: "DEVICE_PRECISE_LOCATION",
    DEVICE_COARSE_LOCATION: "DEVICE_COARSE_LOCATION",
    UPDATE: "UPDATE"
  };

  SurfaceCapabilities = {
    AUDIO_OUTPUT: "actions.capability.AUDIO_OUTPUT",
    SCREEN_OUTPUT: "actions.capability.SCREEN_OUTPUT"
  };
}
