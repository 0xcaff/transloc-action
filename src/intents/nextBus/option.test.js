// @flow
jest.mock("../../now");

import { MockDialogflowApp } from "../../mockDialogflowApp";
import { FROM_ARGUMENT, TO_ARGUMENT } from "./arguments";
import { DialogflowApp } from "actions-on-google";

import { nextBusOption } from "./option";
import { FROM_STOP_KEY } from "./context";

it(`uses the list key for the "to" field`, async () => {
  const app = new MockDialogflowApp();
  app.context.set((FROM_STOP_KEY: any), {
    lifespan: 10,
    name: (FROM_STOP_KEY: any),
    parameters: {
      // Gleason Circle
      stopId: 4197446
    }
  });

  const realApp: DialogflowApp = (app: any);
  app.selectedOption = JSON.stringify({ id: 4209568, type: TO_ARGUMENT });

  await nextBusOption(realApp);

  expect(app.response).toMatchSnapshot();
  expect(app.contextOut).toMatchSnapshot();
});

it(`uses the list key for the "from" field`, async () => {
  const app = new MockDialogflowApp();

  const realApp: DialogflowApp = (app: any);
  app.selectedOption = JSON.stringify({ id: 4197446, type: FROM_ARGUMENT });

  await nextBusOption(realApp);

  expect(app.response).toMatchSnapshot();
  expect(app.contextOut).toMatchSnapshot();
});
