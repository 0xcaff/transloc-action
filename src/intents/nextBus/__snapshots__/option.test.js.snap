// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uses the list key for the "from" field 1`] = `
Object {
  "response": "The following buses are arriving at Gleason Circle. Perkins - Park Point in 11 minutes; Colony Manor / The Province in 16 minutes; RIT Inn in 16 minutes; Racquet Club in 27 minutes; Perkins - Park Point in 32 minutes.",
  "type": "TELL",
}
`;

exports[`uses the list key for the "from" field 2`] = `Map {}`;

exports[`uses the list key for the "to" field 1`] = `
Object {
  "response": "The following buses are traveling from Gleason Circle to Park Point North. Perkins - Park Point in 11 minutes; Perkins - Park Point in 32 minutes.",
  "type": "TELL",
}
`;

exports[`uses the list key for the "to" field 2`] = `Map {}`;
